File: VB/Dennis.Linq/Module.vb
' Developer Express Code Central Example:
' How to populate the list view with data from a LINQ query
' 
' See the http://www.devexpress.com/scid=K18107 KB Article for more information.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E859


Imports Microsoft.VisualBasic
Imports DevExpress.ExpressApp
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.Model.NodeGenerators
Imports DevExpress.ExpressApp.Model.Core
Imports System

Namespace Dennis.Linq
	Public NotInheritable Partial Class LinqModule
		Inherits ModuleBase
		Public Sub New()
			InitializeComponent()
		End Sub
		Public Overrides Sub Setup(ByVal application As XafApplication)
			MyBase.Setup(application)
			AddHandler application.CreateCustomCollectionSource, AddressOf LinqCollectionSourceHelper.CreateCustomCollectionSource
		End Sub
		Public Overrides Sub ExtendModelInterfaces(ByVal extenders As ModelInterfaceExtenders)
			MyBase.ExtendModelInterfaces(extenders)
			extenders.Add(Of IModelListView, IModelListViewLinq)()
		End Sub
		Public Overrides Sub AddGeneratorUpdaters(ByVal updaters As ModelNodesGeneratorUpdaters)
			MyBase.AddGeneratorUpdaters(updaters)
			updaters.Add(New ModelListViewLinqNodesGeneratorUpdater())
			updaters.Add(New ModelListViewLinqColumnsNodesGeneratorUpdater())
		End Sub
	End Class

	Public Interface IModelListViewLinq
	Inherits IModelNode
		Property XPQueryMethod() As String
	End Interface
	Public Class ModelListViewLinqNodesGeneratorUpdater
		Inherits ModelNodesGeneratorUpdater(Of ModelViewsNodesGenerator)
		Public Overrides Sub UpdateNode(ByVal node As ModelNode)
			For Each classInfo As IModelClass In node.Application.BOModel
				If classInfo.TypeInfo.IsPersistent Then
					If (Not String.IsNullOrEmpty(classInfo.Name)) Then
						For Each method As String In LinqCollectionSourceHelper.GetXPQueryMethods(classInfo.TypeInfo.Type)
							Dim id As String = ModelListViewNodesGenerator.GetListViewId(classInfo.TypeInfo.Name) & "_" & method & LinqCollectionSource.DefaultSuffix
							Dim listViewInfo As IModelListView = (TryCast(node.Application.Views.GetNode(id), IModelListView))
							If listViewInfo Is Nothing Then
								listViewInfo = node.AddNode(Of IModelListView)(id)
							End If
							listViewInfo.ModelClass = classInfo
							CType(listViewInfo, IModelListViewLinq).XPQueryMethod = method
						Next method
					End If
				End If
			Next classInfo
		End Sub
	End Class
	Public Class ModelListViewLinqColumnsNodesGeneratorUpdater
		Inherits ModelNodesGeneratorUpdater(Of ModelListViewColumnsNodesGenerator)
		Public Overrides Sub UpdateNode(ByVal node As ModelNode)
			Dim linqViewInfo As IModelListViewLinq = TryCast(node.Parent, IModelListViewLinq)
			If linqViewInfo IsNot Nothing AndAlso (Not String.IsNullOrEmpty(linqViewInfo.XPQueryMethod)) Then
				Dim listViewInfo As IModelListView = CType(linqViewInfo, IModelListView)
				Dim columns() As String = LinqCollectionSourceHelper.GetDisplayableProperties(listViewInfo.ModelClass.TypeInfo.Type, linqViewInfo.XPQueryMethod)
				If columns IsNot Nothing Then
					If listViewInfo.Columns Is Nothing Then
						listViewInfo.AddNode(Of IModelColumns)("Columns")
					End If
					Dim i As Integer = listViewInfo.Columns.Count
					Do While i > 0
						i -= 1
						Dim col As IModelColumn = listViewInfo.Columns.Item(i)
						If Array.IndexOf(columns, col.Id) < 0 Then
						   col.Remove()
						End If
					Loop
					For Each column As String In columns
						Dim col As IModelColumn = (TryCast(listViewInfo.Columns.GetNode(column), IModelColumn))
						If col Is Nothing Then
							col = listViewInfo.Columns.AddNode(Of IModelColumn)(column)
							col.PropertyName = column
						End If
					Next column
				End If
			End If
		End Sub
	End Class
End Namespace

File: CS/Dennis.Linq/Module.cs
// Developer Express Code Central Example:
// How to populate the list view with data from a LINQ query
// 
// See the http://www.devexpress.com/scid=K18107 KB Article for more information.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E859

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.NodeWrappers;
using DevExpress.ExpressApp.InfoGenerators;
using DevExpress.Xpo;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.Model.NodeGenerators;
using DevExpress.ExpressApp.Model.Core;
using System;

namespace Dennis.Linq {
    public sealed partial class LinqModule : ModuleBase {
        public LinqModule() {
            InitializeComponent();
        }
        public override void Setup(XafApplication application) {
            base.Setup(application);
            application.CreateCustomCollectionSource += new System.EventHandler<CreateCustomCollectionSourceEventArgs>(LinqCollectionSourceHelper.CreateCustomCollectionSource);
        }
        public override void ExtendModelInterfaces(ModelInterfaceExtenders extenders) {
            base.ExtendModelInterfaces(extenders);
            extenders.Add<IModelListView, IModelListViewLinq>();
        }
        public override void AddGeneratorUpdaters(ModelNodesGeneratorUpdaters updaters) {
            base.AddGeneratorUpdaters(updaters);
            updaters.Add(new ModelListViewLinqNodesGeneratorUpdater());
            updaters.Add(new ModelListViewLinqColumnsNodesGeneratorUpdater());
        }
    }

    public interface IModelListViewLinq : IModelNode {
        string XPQueryMethod { get; set; }
    }
    public class ModelListViewLinqNodesGeneratorUpdater : ModelNodesGeneratorUpdater<ModelViewsNodesGenerator> {
        public override void UpdateNode(ModelNode node) {
            foreach (IModelClass classInfo in node.Application.BOModel) {
                if (classInfo.TypeInfo.IsPersistent) {
                    if (!string.IsNullOrEmpty(classInfo.Name)) {
                        foreach (string method in LinqCollectionSourceHelper.GetXPQueryMethods(classInfo.TypeInfo.Type)) {
                            string id = ModelListViewNodesGenerator.GetListViewId(classInfo.TypeInfo.Name) + "_" + method + LinqCollectionSource.DefaultSuffix;
                            IModelListView listViewInfo = (node.Application.Views.GetNode(id) as IModelListView);
                            if (listViewInfo == null) {
                                listViewInfo = node.AddNode<IModelListView>(id);
                            }
                            listViewInfo.ModelClass = classInfo;
                            ((IModelListViewLinq)listViewInfo).XPQueryMethod = method;
                        }
                    }
                }
            }
        }
    }
    public class ModelListViewLinqColumnsNodesGeneratorUpdater : ModelNodesGeneratorUpdater<ModelListViewColumnsNodesGenerator> {
        public override void UpdateNode(ModelNode node) {
            IModelListViewLinq linqViewInfo = node.Parent as IModelListViewLinq;
            if (linqViewInfo != null && !string.IsNullOrEmpty(linqViewInfo.XPQueryMethod)) {
                IModelListView listViewInfo = (IModelListView)linqViewInfo;
                string[] columns = LinqCollectionSourceHelper.GetDisplayableProperties(listViewInfo.ModelClass.TypeInfo.Type, linqViewInfo.XPQueryMethod);
                if (columns != null) {
                    if (listViewInfo.Columns == null) {
                        listViewInfo.AddNode<IModelColumns>("Columns");
                    }
                    for (int i = listViewInfo.Columns.Count; i > 0; ) {
                        i--;
                        IModelColumn col = listViewInfo.Columns[i];
                        if (Array.IndexOf(columns, col.Id) < 0) {
                           col.Remove();
                        }
                    }
                    foreach (string column in columns) {
                        IModelColumn col = (listViewInfo.Columns.GetNode(column) as IModelColumn);
                        if (col == null) {
                            col = listViewInfo.Columns.AddNode<IModelColumn>(column);
                            col.PropertyName = column;
                        }
                    }
                }
            }
        }
    }
}